From 9ee20417bd21b0afa887b8e8229da947e519f27c Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:22 -0400
Subject: [net] openvswitch: fix a memory leak

Message-id: <26d019065f89b91ae2a3ec5e4da228545b7f2f8d.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94026
O-Subject: [RHEL7.1 net PATCH 49/50] openvswitch: fix a memory leak
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit 4ee45ea05c8710c7ab8a5eb1a72700b874712746
Author: Li RongQing <roy.qing.li@gmail.com>
Date:   Tue Sep 2 20:52:28 2014 +0800

    openvswitch: fix a memory leak

    The user_skb maybe be leaked if the operation on it failed and codes
    skipped into the label "out:" without calling genlmsg_unicast.

    Cc: Pravin Shelar <pshelar@nicira.com>
    Signed-off-by: Li RongQing <roy.qing.li@gmail.com>
    Acked-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/datapath.c b/net/openvswitch/datapath.c
index 2da06e8..31288fb 100644
--- a/net/openvswitch/datapath.c
+++ b/net/openvswitch/datapath.c
@@ -406,7 +406,7 @@ static int queue_userspace_packet(struct datapath *dp, struct sk_buff *skb,
 {
  struct ovs_header *upcall;
  struct sk_buff *nskb = NULL;
- struct sk_buff *user_skb; /* to be queued to userspace */
+ struct sk_buff *user_skb = NULL; /* to be queued to userspace */
  struct nlattr *nla;
  struct genl_info info = {
   .dst_sk = ovs_dp_get_net(dp)->genl_sock,
@@ -496,9 +496,11 @@ static int queue_userspace_packet(struct datapath *dp, struct sk_buff *skb,
  ((struct nlmsghdr *) user_skb->data)->nlmsg_len = user_skb->len;
 
  err = genlmsg_unicast(ovs_dp_get_net(dp), user_skb, upcall_info->portid);
+ user_skb = NULL;
 out:
  if (err)
   skb_tx_error(skb);
+ kfree_skb(user_skb);
  kfree_skb(nskb);
  return err;
 }
-- 
1.7.1